File: elixir/day5/lib/part1.ex
defmodule Part1 do
  defmodule Graph do
    defstruct edges: %{}, positions_cache: %{}

    def new(rules) do
      edges =
        Enum.reduce(rules, %{}, fn {from, to}, acc ->
          Map.update(acc, from, MapSet.new([to]), &MapSet.put(&1, to))
        end)

      %Graph{edges: edges, positions_cache: %{}}
    end

    def is_valid_sequence(%Graph{} = graph, sequence) do
      positions_cache =
        sequence
        |> Enum.with_index()
        |> Map.new()

      graph = %{graph | positions_cache: positions_cache}

      Enum.all?(sequence, &valid_position?(graph, &1))
    end

    defp valid_position?(graph, page) do
      case Map.get(graph.edges, page) do
        nil -> true
        deps -> Enum.all?(deps, &valid_dependency?(graph, page, &1))
      end
    end

    defp valid_dependency?(graph, page, dep) do
      case Map.get(graph.positions_cache, dep) do
        nil -> true
        dep_pos -> Map.get(graph.positions_cache, page) <= dep_pos
      end
    end
  end

  def find_middle(sequence), do: Enum.at(sequence, div(length(sequence), 2))

  def parse_input(input) do
    [rules_section, updates_section] = String.split(input, "\n\n", trim: true)

    rules =
      rules_section
      |> String.split("\n", trim: true)
      |> Enum.map(&parse_rule/1)

    updates =
      updates_section
      |> String.split("\n", trim: true)
      |> Enum.map(&parse_update/1)

    {rules, updates}
  end

  defp parse_rule(line) do
    [from, to] = String.split(line, "|")
    {String.trim(from) |> String.to_integer(), String.trim(to) |> String.to_integer()}
  end

  defp parse_update(line) do
    line
    |> String.split(",")
    |> Enum.map(&(String.trim(&1) |> String.to_integer()))
  end

  def part1(input) do
    {rules, updates} = parse_input(input)
    graph = Graph.new(rules)

    updates
    |> Enum.filter(&Graph.is_valid_sequence(graph, &1))
    |> Enum.map(&find_middle/1)
    |> Enum.sum()
  end
end

File.read!(~c"input.txt")
|> Part1.part1()
|> IO.inspect(label: "part2")

File: rust/day5/src/bin/part1.rs
use anyhow::{Context, Result};
use std::collections::{HashMap, HashSet};

#[derive(Default)]
struct Graph {
    edges: HashMap<i32, HashSet<i32>>,
    positions_cache: HashMap<i32, usize>,
}

impl Graph {
    fn new(rules: &[(i32, i32)]) -> Self {
        let edges = rules.iter().map(|&(from, to)| (from, to)).fold(
            HashMap::new(),
            |mut acc, (from, to)| {
                acc.entry(from).or_insert_with(HashSet::new).insert(to);
                acc
            },
        );

        Self {
            edges,
            positions_cache: HashMap::with_capacity(32),
        }
    }

    fn is_valid_sequence(&mut self, sequence: &[i32]) -> bool {
        self.positions_cache.clear();
        self.positions_cache.extend(
            sequence
                .iter()
                .copied()
                .enumerate()
                .map(|(i, val)| (val, i)),
        );

        sequence.iter().copied().all(|page| {
            self.edges.get(&page).map_or(true, |deps| {
                deps.iter().all(|&dep| {
                    self.positions_cache
                        .get(&dep)
                        .map_or(true, |&dep_pos| self.positions_cache[&page] <= dep_pos)
                })
            })
        })
    }
}

fn find_middle(sequence: &[i32]) -> Result<i32> {
    if sequence.is_empty() {
        anyhow::bail!("Cannot find middle of empty sequence");
    }
    Ok(sequence[sequence.len() / 2])
}

fn parse_input(input: &str) -> Result<(Vec<(i32, i32)>, Vec<Vec<i32>>)> {
    let (rules_section, updates_section) = input
        .split_once("\n\n")
        .context("Invalid input format: missing section separator")?;

    let rules = rules_section
        .lines()
        .filter(|line| !line.is_empty())
        .enumerate()
        .map(|(line_num, line)| {
            let (from, to) = line.split_once('|').with_context(|| {
                format!(
                    "Invalid rule format at line {}: missing separator",
                    line_num + 1
                )
            })?;

            let from = from.trim().parse().with_context(|| {
                format!("Invalid number '{}' at line {}", from.trim(), line_num + 1)
            })?;
            let to = to.trim().parse().with_context(|| {
                format!("Invalid number '{}' at line {}", to.trim(), line_num + 1)
            })?;

            Ok((from, to))
        })
        .collect::<Result<Vec<_>>>()?;

    let updates = updates_section
        .lines()
        .filter(|line| !line.is_empty())
        .enumerate()
        .map(|(line_num, line)| {
            line.split(',')
                .enumerate()
                .map(|(num_idx, num)| {
                    num.trim().parse::<i32>().with_context(|| {
                        format!(
                            "Invalid number '{}' at position {} in update line {}",
                            num.trim(),
                            num_idx + 1,
                            line_num + 1
                        )
                    })
                })
                .collect::<Result<Vec<_>>>()
        })
        .collect::<Result<Vec<_>>>()?;

    Ok((rules, updates))
}

fn part1(input: &str) -> Result<i32> {
    let (rules, updates) = parse_input(input)?;
    let mut graph = Graph::new(&rules);

    Ok(updates
        .iter()
        .filter(|update| graph.is_valid_sequence(update))
        .map(|update| find_middle(update.as_slice()))
        .collect::<Result<Vec<_>>>()?
        .into_iter()
        .sum())
}

fn main() -> Result<()> {
    let input = include_str!("../../input.txt");
    dbg!(part1(input)?);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        let input = "47|53
97|13
97|61
97|47
75|29
61|13
75|53
29|13
97|29
53|29
61|53
97|53
61|29
47|13
75|47
97|75
47|61
75|61
47|29
75|13
53|13

75,47,61,53,29
97,61,53,29,13
75,29,13
75,97,47,61,53
61,13,29
97,13,75,29,47
";
        assert_eq!(part1(input).unwrap(), 143);
    }
}
